packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,798 @@
+     *
+     * @public
+     */
+    @observable
+    public items: object[] = [];
+    private itemsChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.reset();
+        }
+    }
+
+    /**
+     * The HTML element being used as the viewport
+     *
+     * @beta
+     */
+    @observable
+    public viewportElement: HTMLElement;
+    private viewportElementChanged(): void {
+        if ((this as FoundationElement).$fastController.isConnected) {
+            this.resetAutoUpdateMode(this.autoUpdateMode, this.autoUpdateMode);
+        }
+    }
+
+    /**
+     * The ViewTemplate used to render items.
+     *
+     * @public
+     */
+    @observable
+    public itemTemplate: ViewTemplate = defaultItemTemplate;
+
+    /**
+     * The items that are currently visible (includes buffer regions)
+     *
+     * @internal
+     */
+    @observable
+    public visibleItems: any[] = [];
+
+    /**
+     * The positions of the currently rendered items in the list
+     *
+     * @internal
+     */
+    @observable
+    public visibleItemSpans: SpanMap[] = [];
+
+    /**
+     * The calculated span of the total stack.
+     * (ie. all items + start/end regions)
+     *
+     * @internal
+     */
+    @observable
+    public totalListSpan: number = 0;
+
+    /**
+     * The size in pixels of the start "spacer"
+     * (ie. the grid region that holds space for non-rendered elements at the start of the stack)
+     *
+     * @internal
+     */
+    @observable
+    public startSpacerSpan: number = 0;
+
+    /**
+     * The size in pixels of the end "spacer"
+     * (ie. the grid region that holds space for non-rendered elements at the end of the stack)
+     *
+     * @internal
+     */
+    @observable
+    public endSpacerSpan: number = 0;
+
+    /**
+     * The index of the first item in the array to be rendered
+     *
+     * @internal
+     */
+    @observable
+    public firstRenderedIndex: number = -1;
+
+    /**
+     * The index of the last item in the array to be rendered
+     *
+     * @internal
+     */
+    @observable
+    public lastRenderedIndex: number = -1;
+
+    /**
+     * reference to the container element
+     *
+     * @internal
+     */
+    public containerElement: HTMLDivElement;
+
+    private static intersectionService: IntersectionService = new IntersectionService();
+    private resizeDetector: ResizeObserverClassDefinition | null = null;
+
+    private pendingPositioningUpdate: boolean = false;
+    private pendingReset: boolean = false;
+
+    private visibleRangeStart: number = 0;
+    private visibleRangeEnd: number = 0;
+
+    private viewportRect: ClientRect | DOMRect | undefined;
+    private containerRect: ClientRect | DOMRect | undefined;
+
+    private itemsRepeatBehavior: RepeatBehavior | null = null;
+    private itemsPlaceholder: Node;
+
+    private itemsObserver: Notifier | null = null;
+    private itemCount: number = 0;
+
+    private finalUpdate: boolean = false;
```

Comment:
There are a lot of private variables here - some doc blocks would be useful to describe what they are and how they're used.

Reply:
Did some cleanup around these.